modules/virtual_network_gateway/variables.tf

```diff
@@ -0,0 +1,324 @@
+variable "resource_group_name" {
+  description = "Name of a pre-existing Resource Group to place the resources in."
+  type        = string
+}
+
+variable "location" {
+  description = "Region to deploy load balancer and dependencies."
+  type        = string
+}
+
+variable "name_prefix" {
+  description = "A prefix added to all resource names created by this module"
+  default     = ""
+  type        = string
+}
+
+variable "name_suffix" {
+  description = "A suffix added to all resource names created by this module"
+  default     = ""
+  type        = string
+}
+
+variable "name" {
+  description = "The name of the Virtual Network Gateway. Changing this forces a new resource to be created"
+  type        = string
+}
+
+variable "tags" {
+  description = "Azure tags to apply to the created resources."
+  default     = {}
+  type        = map(string)
+}
+
+variable "enable_zones" {
+  description = "If false, all the subnet-associated frontends and also all created Public IP addresses default to not to use Availability Zones (the `No-Zone` setting). It is intended for the regions that do not yet support Availability Zones."
+  default     = true
+}
+
+variable "avzones" {
+  description = <<-EOF
+  After provider version 3.x you need to specify in which availability zone(s) you want to place IP.
+  ie: for zone-redundant with 3 availability zone in current region value will be:
+  ```["1","2","3"]```
+  EOF
+  default     = []
+  type        = list(string)
+}
+
+variable "type" {
+  description = "The type of the Virtual Network Gateway. Valid options are Vpn or ExpressRoute. Changing the type forces a new resource to be created"
+  type        = string
+}
+
+variable "vpn_type" {
+  description = "The routing type of the Virtual Network Gateway. Valid options are RouteBased or PolicyBased. Defaults to RouteBased. Changing this forces a new resource to be created."
+  default     = "RouteBased"
+  type        = string
+}
+
+variable "sku" {
```

Comment:
This depends on Generation and SKU, should we add some validation? Or just rely on error messages thrown by the API?

Reply:
Thank you for feedback, you are right.

I've added validation for `type`, `vpn_type`, `sku` and `generation` e.g.: https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/64b2fbfce7645aa94d04ec23e2c2a11f8be6e2b1/modules/virtual_network_gateway/variables.tf#L99-L107

Reply:
i was not clear enough, I thought about adding precondition block

Reply:
@FoSix , no problem, thank you for feedback.

I've added `lifecycle precondition` to check SKU, which depends on generation: https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/872dce18522fce79418298c412c036ccf8e34c7b/modules/virtual_network_gateway/main.tf#L80C3-L85

I hope it fullfil requirement about check generation and SKU.